File: functions/javascript/src/handlePlayNotification.ts
/**
 * handlePlayNotification.ts — Google Play RTDN（リアルタイム開発者通知）ハンドラ
 *
 * Google Play Console で設定した Cloud Pub/Sub テーマに配信される
 * サブスクリプション通知（RTDN: Real-time Developer Notifications）を受信し、
 * Google Play Developer API で最新のサブスクリプション状態を再検証したうえで、
 * Firestore のユーザーデータを更新する Cloud Function。
 *
 * 【処理フロー】
 * 1. Pub/Sub メッセージから RTDN データをデコード
 * 2. subscriptionNotification が含まれていない場合はスキップ（テスト通知など）
 * 3. purchaseToken で Firestore からユーザーを検索
 * 4. Google Play Developer API v3 で購入状態を再検証
 * 5. 検証結果に基づいて tier と subscription 情報を Firestore に保存
 *
 * 【RTDN の notificationType 一覧（一部）】
 * - 1: 復旧（RECOVERED）
 * - 2: 更新（RENEWED）
 * - 3: キャンセル（CANCELED）
 * - 4: 購入（PURCHASED）
 * - 12: 失効（EXPIRED）
 * - 13: 期限切れ（EXPIRED）
 * ※ 本関数では notificationType に関わらず Play API で再検証するため、
 *   個別の通知タイプごとの処理は行わない
 *
 * Pub/Sub テーマ: play-subscription-notifications
 * リージョン: asia-northeast1（東京）
 */
import * as functions from 'firebase-functions/v2';
import * as admin from 'firebase-admin';
import { verifyPlayPurchase } from './services/playBilling';
import {
  FREE_DAILY_SENTENCES, FREE_DAILY_QUIZZES,
  PREMIUM_DAILY_SENTENCES, PREMIUM_DAILY_QUIZZES,
} from './constants/quota';

/** Firestore インスタンス */
const db = admin.firestore();

/**
 * RTDN（Real-time Developer Notifications）メッセージの型定義
 *
 * Google Play から Pub/Sub 経由で送信される通知メッセージの構造。
 * subscriptionNotification が含まれない場合はテスト通知等のため処理をスキップする。
 */
interface RTDNMessage {
  /** サブスクリプション通知データ（テスト通知の場合は undefined） */
  subscriptionNotification?: {
    /** RTDN バージョン */
    version: string;
    /** 通知タイプ（1=復旧, 2=更新, 3=キャンセル, 4=購入, 12=失効, 13=期限切れ など） */
    notificationType: number;
    /** Google Play の購入トークン（ユーザー検索に使用） */
    purchaseToken: string;
    /** サブスクリプション ID（商品 ID） */
    subscriptionId: string;
  };
  /** アプリのパッケージ名 */
  packageName: string;
}

/**
 * handlePlayNotification - Google Play RTDN Pub/Sub ハンドラ
 *
 * Google Play から Pub/Sub 経由でサブスクリプション通知を受け取り、
 * Play Developer API で再検証後、Firestore を更新する。
 */
export const handlePlayNotification = functions.pubsub.onMessagePublished(
  {
    topic: 'play-subscription-notifications',
    region: 'asia-northeast1',
  },
  async (event) => {
    // Pub/Sub メッセージのデータ部分を Base64 デコードして JSON パース
    const messageData = event.data.message.data
      ? JSON.parse(Buffer.from(event.data.message.data, 'base64').toString()) as RTDNMessage
      : null;

    // subscriptionNotification がない場合はテスト通知等のためスキップ
    if (!messageData?.subscriptionNotification) {
      console.log('Non-subscription notification, skipping');
      return;
    }

    const { notificationType, purchaseToken, subscriptionId } =
      messageData.subscriptionNotification;
    const packageName = messageData.packageName;

    console.log(
      `RTDN: type=${notificationType}, subscriptionId=${subscriptionId}`
    );

    // purchaseToken で Firestore からユーザーを検索
    // verifySubscription で保存した purchase_token と照合する
    // 匿名ユーザーの再インストール等で同一サブスクの doc が複数残る可能性が
    // あるため limit(1) にせず、該当する全 doc を更新する
    const usersSnapshot = await db
      .collection('users')
      .where('subscription.purchase_token', '==', purchaseToken)
      .get();

    // 該当ユーザーが見つからない場合（初回購入前の通知など）
    if (usersSnapshot.empty) {
      console.warn('No user found for purchaseToken');
      return;
    }

    // Google Play Developer API v3 で最新のサブスクリプション状態を取得
    try {
      const result = await verifyPlayPurchase(
        packageName,
        subscriptionId,
        purchaseToken
      );

      // 検証結果に基づいて tier を決定（expired なら free、それ以外は premium）
      const tier = result.status === 'expired' ? 'free' : 'premium';
      const isFree = tier === 'free';

      for (const userDoc of usersSnapshot.docs) {
        const tierChanged = userDoc.data()?.tier !== tier;

        // Firestore のユーザードキュメントを更新
        // ドット記法で subscription のサブフィールドのみ更新し、purchase_token 等を保持する
        const updateData: Record<string, unknown> = {
          tier,
          'subscription.status': result.status,
          'subscription.expires_at': result.expiresAt
            ? admin.firestore.Timestamp.fromDate(result.expiresAt)
            : null,
          'subscription.auto_renewing': result.autoRenewing,
          'subscription.updated_at': admin.firestore.FieldValue.serverTimestamp(),
        };

        // クォータはティアが変わる時のみリセット（更新通知で誤リセットしない）
        if (tierChanged) {
          updateData.remaining_sentences = isFree ? FREE_DAILY_SENTENCES : PREMIUM_DAILY_SENTENCES;
          updateData.remaining_quizzes = isFree ? FREE_DAILY_QUIZZES : PREMIUM_DAILY_QUIZZES;
        }

        await userDoc.ref.update(updateData);

        console.log(
          `Updated user ${userDoc.id}: tier=${tier}, status=${result.status}`
        );
      }
    } catch (error) {
      console.error('Failed to verify Play purchase on RTDN:', error);
    }
  }
);

File: functions/go/handle_play_notification.go
package function

import (
	"context"
	"encoding/json"
	"log"

	"cloud.google.com/go/firestore"
	"github.com/cloudevents/sdk-go/v2/event"
	"google.golang.org/api/iterator"

	"github.com/mnbst/thai-memo/functions/go/internal/fbapp"
	"github.com/mnbst/thai-memo/functions/go/internal/playbilling"
	"github.com/mnbst/thai-memo/functions/go/internal/quota"
)

// handlePlayNotification は
// functions/javascript/src/handlePlayNotification.ts の移植。
//
// Google Play Console で設定した Cloud Pub/Sub テーマに配信される
// サブスクリプション通知（RTDN: Real-time Developer Notifications）を受信し、
// Google Play Developer API で最新のサブスクリプション状態を再検証したうえで、
// Firestore のユーザーデータを更新する。
//
// notificationType に関わらず Play API で再検証するため、
// 個別の通知タイプごとの処理は行わない。

// messagePublishedData は Pub/Sub の CloudEvent データ。
type messagePublishedData struct {
	Message struct {
		// Data は base64 で載ってくる。[]byte は encoding/json が自動で戻す。
		Data       []byte            `json:"data"`
		Attributes map[string]string `json:"attributes"`
		MessageID  string            `json:"messageId"`
	} `json:"message"`
	Subscription string `json:"subscription"`
}

// rtdnMessage は Google Play から Pub/Sub 経由で届く通知メッセージ。
// subscriptionNotification が無い場合はテスト通知等のためスキップする。
type rtdnMessage struct {
	SubscriptionNotification *struct {
		Version          string `json:"version"`
		NotificationType int    `json:"notificationType"`
		// PurchaseToken はユーザー検索に使う
		PurchaseToken  string `json:"purchaseToken"`
		SubscriptionID string `json:"subscriptionId"`
	} `json:"subscriptionNotification"`
	PackageName string `json:"packageName"`
}

func handlePlayNotificationEvent(ctx context.Context, e event.Event) error {
	var data messagePublishedData
	if err := e.DataAs(&data); err != nil {
		return err
	}

	var message rtdnMessage
	if len(data.Message.Data) > 0 {
		if err := json.Unmarshal(data.Message.Data, &message); err != nil {
			return err
		}
	}

	if message.SubscriptionNotification == nil {
		log.Print("Non-subscription notification, skipping")
		return nil
	}
	notification := message.SubscriptionNotification

	log.Printf("RTDN: type=%d, subscriptionId=%s",
		notification.NotificationType, notification.SubscriptionID)

	return processPlayNotification(ctx, message.PackageName,
		notification.SubscriptionID, notification.PurchaseToken)
}

func processPlayNotification(
	ctx context.Context, packageName, subscriptionID, purchaseToken string,
) error {
	db, err := fbapp.Firestore(ctx)
	if err != nil {
		return err
	}

	// purchaseToken で Firestore からユーザーを検索。
	// verifySubscription で保存した purchase_token と照合する。
	// 匿名ユーザーの再インストール等で同一サブスクの doc が複数残る可能性が
	// あるため limit(1) にせず、該当する全 doc を更新する。
	it := db.Collection("users").
		Where("subscription.purchase_token", "==", purchaseToken).
		Documents(ctx)
	defer it.Stop()

	var docs []*firestore.DocumentSnapshot
	for {
		doc, err := it.Next()
		if err == iterator.Done {
			break
		}
		if err != nil {
			return err
		}
		docs = append(docs, doc)
	}

	if len(docs) == 0 {
		// 初回購入前の通知など
		log.Print("No user found for purchaseToken")
		return nil
	}

	result, err := playbilling.Default.VerifyPurchase(
		ctx, packageName, subscriptionID, purchaseToken)
	if err != nil {
		// JS 版は再検証の失敗をログに落として握り潰す（Pub/Sub の再送を招かない）。
		log.Printf("Failed to verify Play purchase on RTDN: %v", err)
		return nil
	}

	// 検証結果に基づいて tier を決定（expired なら free、それ以外は premium）
	tier := "premium"
	if result.Status == playbilling.StatusExpired {
		tier = "free"
	}

	for _, doc := range docs {
		currentTier, _ := doc.Data()["tier"].(string)
		if _, err := doc.Ref.Update(ctx, playUpdates(result, tier, currentTier)); err != nil {
			return err
		}
		log.Printf("Updated user %s: tier=%s, status=%s",
			doc.Ref.ID, tier, result.Status)
	}
	return nil
}

// playUpdates は1ユーザーぶんの更新内容を組み立てる（Firestore に触らない）。
//
// ドット記法で subscription のサブフィールドのみ更新し、purchase_token 等を保持する。
func playUpdates(
	result *playbilling.VerificationResult, tier, currentTier string,
) []firestore.Update {
	isFree := tier == "free"

	updates := []firestore.Update{
		{Path: "tier", Value: tier},
		{Path: "subscription.status", Value: string(result.Status)},
		{Path: "subscription.auto_renewing", Value: result.AutoRenewing},
		{Path: "subscription.updated_at", Value: firestore.ServerTimestamp},
	}
	if result.ExpiresAt != nil {
		updates = append(updates, firestore.Update{
			Path: "subscription.expires_at", Value: *result.ExpiresAt,
		})
	} else {
		updates = append(updates, firestore.Update{
			Path: "subscription.expires_at", Value: nil,
		})
	}

	// クォータはティアが変わる時のみリセット（更新通知で誤リセットしない）
	if currentTier != tier {
		sentences, quizzes := quota.PremiumDailySentences, quota.PremiumDailyQuizzes
		if isFree {
			sentences, quizzes = quota.FreeDailySentences, quota.FreeDailyQuizzes
		}
		updates = append(updates,
			firestore.Update{Path: "remaining_sentences", Value: sentences},
			firestore.Update{Path: "remaining_quizzes", Value: quizzes},
		)
	}

	return updates
}
